migrator.go

```diff
@@ -0,0 +1,84 @@
+package influxdb
+
+import "context"
+
+// MigrationManager is the manager of migrators.
+type MigrationManager struct {
+	Migrators []Migrator
+}
+
+// Migrate will perform the migration.
+func (mg MigrationManager) Migrate(ctx context.Context, Up bool, errCh chan<- error) {
+	for _, m := range mg.Migrators {
+		if !m.NeedMigration() {
+			return
+		}
+		bucket := m.Bucket()
+		srcs, err := m.ScanBytes(ctx, bucket)
+		if err != nil {
+			errCh <- &Error{
+				Err: err,
+				Op:  "Data Migration Scan",
+			}
+			continue
+		}
+		for key, src := range srcs {
+			var dst []byte
+			if Up {
+				dst, err = m.Up(src)
+			} else {
+				dst, err = m.Down(src)
+			}
+			if err != nil {
+				errCh <- err
+				continue
+			}
+			if err = m.StoreBytes(ctx, bucket, key, dst); err != nil {
+				errCh <- err
+			}
+		}
+	}
+	// close it
+	errCh <- nil
+}
+
+// Migrator is the interface of schema migration.
+type Migrator interface {
+	MigragorService
+	// Bucket is the bucket of the data stored.
+	Bucket() []byte
+	// NeedMigration will return whether this migrator is needed at runtime.
+	NeedMigration() bool
+	// Up convert the old version to the new version.
+	Up(src []byte) (dst []byte, err error)
+	// Down convert the new version to the old version.
+	Down(src []byte) (dst []byte, err error)
+}
+
+// MigragorService is the service to fetch and store bytes.
+type MigragorService interface {
+	ScanBytes(ctx context.Context, bucket []byte) (data map[string][]byte, err error)
```

Comment:
Any attempt to read all key's and values from an entire bucket will cause the system to run out of memory.

Reply:
Good catch